lib/java/js-websocket/src/main/java/org/enso/polyfill/timers/TimersPolyfill.java

```diff
@@ -0,0 +1,103 @@
+package org.enso.polyfill.timers;
+
+import java.util.Arrays;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+import org.enso.polyfill.Polyfill;
+import org.graalvm.polyglot.Context;
+import org.graalvm.polyglot.Source;
+import org.graalvm.polyglot.Value;
+import org.graalvm.polyglot.proxy.ProxyExecutable;
+
+public final class TimersPolyfill implements ProxyExecutable, Polyfill {
+
+    private static final String SET_INTERVAL = "set-interval";
+    private static final String CLEAR_INTERVAL = "clear-interval";
+    private static final String SET_TIMEOUT = "set-timeout";
+    private static final String CLEAR_TIMEOUT = "clear-timeout";
+
+    private static final String TIMERS_POLYFILL_JS = "timers-polyfill.js";
+
+    private static final TimeUnit TIME_UNIT = TimeUnit.MILLISECONDS;
+
+    private final ScheduledExecutorService scheduledExecutor = Executors.newSingleThreadScheduledExecutor((r) -> {
+        var thread = new Thread(r);
+        thread.setDaemon(true);
+        return thread;
```

Comment:
I'm not super sure about how this executor works, but it looks like it creates a single, but separate thread for its own execution. On the surface it looks incorrect, since JS environments always assume fully single-threaded execution per context. That means ALL calls into JS should be evaluated on the same thread. There should never be the case where a timeout handler is executed truly concurrently with e.g. a websocket data handler. All handlers need to be fully serialized onto a single thread. Is this maybe already happening somewhere else?

Reply:
> never ... timeout handler is executed truly concurrently with e.g. a websocket data handler

That's guaranteed by [Graal.js](https://github.com/oracle/graaljs) own implementation. Should there be a violation of the _single threadedness_ and a concurrent access is made, an exception is yielded. In fact we have problems to overcome and customize that behavior and  we have [proposed some improvements](https://github.com/oracle/graal/pull/8266).

[node.js has event queue](https://nodejs.org/en/learn/asynchronous-work/event-loop-timers-and-nexttick). Using this dedicated thread is our way to simulate such _event queue_. The idea is that all JavaScript execution will be happening in this only thread.